android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt

```diff
@@ -219,4 +208,93 @@ constructor(
 
   suspend fun getTask(id: String) =
     kotlin.runCatching { fhirEngine.get<Task>(id) }.onFailure { Timber.e(it) }.getOrNull()
+
+  private fun evaluateToDate(base: Base?, expression: String): BaseDateTimeType? =
+    base?.let { fhirPathEngine.evaluate(it, expression).firstOrNull()?.dateTimeValue() }
+
+  private fun PlanDefinition.PlanDefinitionActionComponent.passesConditions(
+    focus: Resource?,
+    root: Resource?,
+    base: Base
+  ) =
+    this.condition.all {
+      require(it.kind == PlanDefinition.ActionConditionKind.APPLICABILITY) {
+        "PlanDefinition.action.kind=${it.kind} not supported"
+      }
+
+      require(it.expression.language == Expression.ExpressionLanguage.TEXT_FHIRPATH.toCode()) {
+        "PlanDefinition.expression.language=${it.expression.language} not supported"
+      }
+
+      fhirPathEngine.evaluateToBoolean(focus, root, base, it.expression.expression)
+    }
+
+  private fun PlanDefinition.PlanDefinitionActionComponent.activityDefinition(
+    planDefinition: PlanDefinition
+  ) =
+    planDefinition.contained.filter { it.resourceType == ResourceType.ActivityDefinition }.first {
+      it.logicalId == this.definitionCanonicalType.value
+    } as
+      ActivityDefinition
+
+  private fun PlanDefinition.PlanDefinitionActionComponent.taskPeriods(
+    definition: ActivityDefinition,
+    carePlan: CarePlan
+  ): List<Period> {
+    return when {
+      definition.hasDosage() -> extractTaskPeriodsFromDosage(definition.dosage, carePlan)
+      definition.hasTiming() && !definition.hasTimingTiming() ->
+        throw IllegalArgumentException(
+          "Timing component should only be Timing. Can not handle ${timing.fhirType()}"
+        )
+      else -> extractTaskPeriodsFromTiming(definition.timingTiming, carePlan)
+    }
+  }
+
+  private fun extractTaskPeriodsFromTiming(timing: Timing, carePlan: CarePlan): List<Period> {
+    val taskPeriods = mutableListOf<Period>()
+    // TODO handle properties used by older plandefintions. If any plan definition is using
+    //  countMax, frequency, durationUnit of hour, consider as non compliant and assume
+    //  all handling would be done into structure map. Once all plan definitions are using
+    //  recommended approach and using plan definitions properly change line below to use
+    //  timing.repeat.count only
+    val isLegacyPlanDefinition =
+      (timing.repeat.hasFrequency() ||
+        timing.repeat.hasCountMax() ||
+        timing.repeat.durationUnit?.equals(UnitsOfTime.H) == true)
+    val count = if (isLegacyPlanDefinition || !timing.repeat.hasCount()) 1 else timing.repeat.count
+    val periodExpression = timing.extractFhirpathPeriod()
+    val durationExpression = timing.extractFhirpathDuration()
+
+    // offset date for current task period; careplan start if all tasks generated at once
+    // otherwise today means that tasks are generated on demand
```

Comment:
```suggestion
    // Offset date for current task period; CarePlan start if all tasks generated at once
    // otherwise today means that tasks are generated on demand
```